ports.go

```diff
@@ -26,140 +28,105 @@ type DevicePortServiceOp struct {
-	device, _, err := i.client.Devices.GetExtra(input.DeviceID, []string{"virtual_networks"}, nil)
+type DisbondRequest struct {
+	PortID      string `json:"id"`
+	BulkDisable bool   `json:"bulk_disable"`
+}
 
+func (i *DevicePortServiceOp) GetBondedPort(deviceID string) (*Port, error) {
+	device, _, err := i.client.Devices.Get(deviceID)
+	if err != nil {
+		return nil, err
+	}
 	for _, port := range device.NetworkPorts {
-		if port.ID == input.PortID && !port.hasVirtualNetwork(input.VirtualNetworkID) {
-			if len(port.AttachedVirtualNetworks) == 0 {
-				// convert to layer-2 (and attach VLAN)
-				return i.convertToLayerTwo(input)
-			} else {
-				// not the first VLAN, so attach without converting
-				return i.assignVirtualNetwork(input)
-			}
+		if port.Type == "NetworkBondPort" {
+			return &port, nil
 		}
 	}
 
-	return nil, nil, false, err
-}
-
-type PortUnassignRequest struct {
-	DeviceID         string
-	PortID           string
-	VirtualNetworkID int `json:"vnid"`
-}
-
-type PortUnassignResponse struct {
-	PortID          string           `json:"id"`
-	VirtualNetworks []VirtualNetwork `json:"virtual_networks"`
+	return nil, fmt.Errorf("No bonded port found in device %s", deviceID)
 }
 
-func (i *DevicePortServiceOp) Unassign(input *PortUnassignRequest) (*PortUnassignResponse, *Response, bool, error) {
-	path := fmt.Sprintf("%s/%s/unassign", portBasePath, input.PortID)
-	unassignResponse := new(PortUnassignResponse)
-
-	resp, err := i.client.DoRequest("POST", path, input, unassignResponse)
+func (i *DevicePortServiceOp) GetPortByName(deviceID, name string) (*Port, error) {
+	device, _, err := i.client.Devices.Get(deviceID)
 	if err != nil {
-		return nil, resp, false, err
+		return nil, err
+	}
+	for _, port := range device.NetworkPorts {
+		if port.Name == name {
+			return &port, nil
+		}
 	}
 
-	return unassignResponse, resp, true, err
+	return nil, fmt.Errorf("Port %s not found in device %s", name, deviceID)
 }
 
-type PortBondRequest struct {
-	PortID     string
-	BulkEnable bool
+func (i *DevicePortServiceOp) Assign(portID, vlanID string) (*Port, *Response, error) {
+	path := fmt.Sprintf("%s/%s/assign", portBasePath, portID)
+	return i.assignmentAction(portID, vlanID, path)
 }
 
-func (i *DevicePortServiceOp) Bond(portID string, bulkEnable bool) (*Port, *Response, error) {
-	path := fmt.Sprintf("%s/%s/bond", portBasePath, portID)
-	if bulkEnable {
-		path += "?bulk_enable=true"
-	}
-	output := new(Port)
-
-	resp, err := i.client.DoRequest("POST", path, nil, output)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	return output, resp, nil
+func (i *DevicePortServiceOp) Unassign(portID, vlanID string) (*Port, *Response, error) {
+	path := fmt.Sprintf("%s/%s/unassign", portBasePath, portID)
+	return i.assignmentAction(portID, vlanID, path)
 }
 
-type PortDisbondRequest struct {
-	PortID      string
-	BulkDisable bool
+func (i *DevicePortServiceOp) assignmentAction(portID, vlanID, path string) (*Port, *Response, error) {
```

Comment:
I'm not a fan of using this for both `Assign` and `Unassign` in the event that should you ever need to change the logic of how one of these two functions work then it makes it difficult since they follow the same logic path

Reply:
I got rid of the assignmentAction and kept only portAction. It really saves space to do the API request for these 4 similar tasks in one method, and it's kept readable I think.